assets/src/stories-editor/components/higher-order/with-call-to-action-validation.js

```diff
@@ -1,61 +1,64 @@
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+
 /**
  * WordPress dependencies
  */
 import { getBlockType } from '@wordpress/blocks';
 import { Button } from '@wordpress/components';
-import { withSelect, withDispatch } from '@wordpress/data';
+import { useSelect, useDispatch } from '@wordpress/data';
 import { Warning } from '@wordpress/block-editor';
-import { createHigherOrderComponent, compose } from '@wordpress/compose';
+import { createHigherOrderComponent } from '@wordpress/compose';
 import { __ } from '@wordpress/i18n';
+import { useCallback } from '@wordpress/element';
 
 /**
  * Internal dependencies
  */
 import { isCTABlock } from '../../helpers';
 
-const enhance = compose(
-	/**
-	 * For blocks which are only allowed once per page,  provides the
-	 * wrapped component with `originalBlockClientId` -- a reference to the
-	 * first block of the same type on the page -- if and only if that
-	 * "original" block is not the current one. Thus, an non-existent
-	 * `originalBlockClientId` prop signals that the block is valid.
-	 */
-	withSelect( ( select, props ) => {
-		const { getBlockRootClientId, getBlock, getBlockOrder, getBlocksByClientId } = select( 'core/block-editor' );
-
-		if ( ! isCTABlock( props.name ) ) {
-			return {};
-		}
+export default createHigherOrderComponent( ( BlockEdit ) => {
+	const CallToActionValidation = ( props ) => {
+		const {
+			isInvalid,
+			originalBlockClientId,
+		} = useSelect( ( select ) => {
+			const { getBlockRootClientId, getBlock, getBlockOrder, getBlocksByClientId } = select( 'core/block-editor' );
 
-		const parentBlock = getBlock( getBlockRootClientId( props.clientId ) );
+			if ( ! isCTABlock( props.name ) ) {
+				return {};
+			}
 
-		if ( ! parentBlock ) {
-			return {};
-		}
+			const parentBlock = getBlock( getBlockRootClientId( props.clientId ) );
 
-		const isFirstPage = getBlockOrder().indexOf( parentBlock.clientId ) === 0;
-		const blocksOnPage = getBlocksByClientId( getBlockOrder( parentBlock.clientId ) );
-		const firstOfSameType = blocksOnPage.find( ( { name } ) => name === props.name );
-		const isInvalid = firstOfSameType && firstOfSameType.clientId !== props.clientId;
-
-		return {
-			isInvalid: isFirstPage || isInvalid,
-			originalBlockClientId: isInvalid && firstOfSameType.clientId,
-		};
-	} ),
-	withDispatch( ( dispatch, { originalBlockClientId } ) => ( {
-		selectFirst: () => dispatch( 'core/block-editor' ).selectBlock( originalBlockClientId ),
-	} ) ),
-);
+			if ( ! parentBlock ) {
+				return {};
+			}
+
+			const isFirstPage = getBlockOrder().indexOf( parentBlock.clientId ) === 0;
+			const blocksOnPage = getBlocksByClientId( getBlockOrder( parentBlock.clientId ) );
```

Comment:
There is a helper `getCallToActionBlock`, that can be used here for convenience.

Reply:
You mean `getPageBlockByName`? Shouldn't that be a hook then?

Reply:
Well, yes is should. We should come up with a good strategy for `select`'ing helpers, that span multiple registries. Because it's not a selector (they can't be multi registry), and it's not a plain function. And if we change it to a hook, it **has** to be called on every render - and this one won't as there's some conditional returns above.

Reply:
Why is `getCallToActionBlock` useful here? This gets a list of blocks on page..

Reply:
And on the next line that list is filtered by block type, just like `getCallToActionBlock` does.